qutip/core/qobj.py

```diff
@@ -833,25 +850,25 @@ def diag(self):
             Returns array of ``real`` values if operators is Hermitian,
             otherwise ``complex`` values are returned.
         """
-        out = self.data.as_scipy().diagonal()
+        # TODO: add a `diagonal` method to the data layer?
+        out = _data.to(_data.CSR, self.data).as_scipy().diagonal()
         if np.any(np.imag(out) > settings.core['atol']) or not self.isherm:
             return out
         else:
             return np.real(out)
 
     @_tidyup
-    def expm(self, method='dense'):
+    def expm(self, dtype=_data.Dense):
```

Comment:
Why ask for the data type here, instead of the string? This force the user to know where to find the data layer object, which I don't like... But the string are messy and cannot be extended as easily to new types... I don't like it but it has it's advantages.

Reply:
Yeah, I don't like this either.  It's a similar problem with `eigs`, but that's even less obvious.  I'm going to think about it a bit more, and hopefully I'll have some more ideas to suggest - it's one of my to do bullet points.